doc: Updated nRF54H20 power management doc (#23275)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54h/ug_nrf54h20_architecture_pm.rst

For reference implementations, see the :ref:`multicore_idle_test` samples.
These samples demonstrate configurations where the radio core remains idle while the application core remains active.

Software power states
*********************

The nRF54H20 ARM Cortex CPUs currently support the following software power states:

* Active
* Idle

> * Idle with cache disabled

Reviewer: We need to describe details of this state in a new subsection below

The power consumption in this state depends on the overall System state but is lower than in any of the *Idle* states.
The wake-up latency is higher than in any of the *Idle* states due to the CPU state restoration procedure.

Mapping application states to software power states
***************************************************

> The following table maps application states to software power states:

Author: @hubertmis  Please check if the mapping in the table is correct

Reviewer: Yes, I think that's good. It doesn't show all the details, but adding them would make it hard to read

| Application state   | Application core power state | Radio core power state                | SoC hardware state                       |
+=====================+==============================+=======================================+==========================================+
| Active              | Active,                      | Active,                               | System On,                               |
|                     | Idle,                        | Idle,                                 | System On Idle,                          |
|                     | Local suspend to RAM         | Idle with cache disabled              | System On All Idle                       |
+---------------------+------------------------------+---------------------------------------+------------------------------------------+
| Idle                | Local suspend to RAM         | Idle with cache disabled              | System On All Idle                       |
+---------------------+------------------------------+---------------------------------------+------------------------------------------+

Power management optimization
*****************************

> For suggestions on power management optimization techniques on the nRF54H20 SoC, see the :ref:`ug_nrf54h20_pm_optimization` page.

Reviewer 2: ```suggestion
For recommendations on power management optimization techniques on the nRF54H20 SoC, see the :ref:`ug_nrf54h20_pm_optimization` page.
```

**************************

To benchmark the power consumption in *Idle* state, see :ref:`multicore_idle_test`.

Power management optimization example
*************************************

> This example focuses on a HID device use case (for example, a mouse)

Reviewer: I would start that the mouse has two main application states:

1. Active - when the user is moving the mouse
2. Idle - after an inactivity timeout, until the user picks up the device again

The workflow below applies to the "Active" application state.

In this use case the typical workflow is the following:

  1. Sample input data from sensors.
  2. Construct a HID report.
  3. Transmit the report over the radio.
  4. Sleep until the next period (usually, less than 1 ms).

The power consumption targets for this example are the following:

* Sleep currents (5 V supply):

  * Application core (Suspend-to-RAM): < 10 µA

>   * Radio core (Suspend-to-Idle, cache disabled): < 10 µA

Reviewer: This is the mapping between Application states and Software power states:

1. In Application Active state:
    1. Application CPU can be in:
       1. Active
       2. Idle
       3. Suspend-to-RAM
    2. Radio CPU can be in:
       1. Active
       2. Idle
       3. Idle with cache disabled
    3. With this combination of software power states, depending on the current states, the hardware switches between:
       1. System On
       2. System On Idle
       3. System On All Idle

4. In the Idle application state:
   1. Application CPU is in Suspend-to-RAM
   2. Radio CPU is in Idle with cache disabled
   5. With this combination of software power states, the hardware is in System On All Idle

* Active currents (estimates at 5 V):

>   * Radio core in WFI: ~0.25 mA
>   * App core in WFI: ~0.15 mA
>   * Core wake-up every 1 ms: +0.05 mA
>   * ADC sample every 1 ms: +0.10 mA
>   * SPI transaction every 1 ms: +0.20 mA
>   * IPC message every 1 ms: +0.15 mA

Reviewer: To unify terminology (WFI is an instruction that suspends the CPU, which leads to the Idle power state).

```suggestion
  * Radio core in Idle: ~0.25 mA
  * App core in Idle: ~0.15 mA
  * Core wake-up every 1 ms: +0.05 mA
  * ADC sample every 1 ms: +0.10 mA
  * SPI transaction every 1 ms: +0.20 mA
  * IPC message every 1 ms: +0.15 mA
```

  * CONFIG_PM_S2RAM=y
  * CONFIG_POWEROFF=y
  * CONFIG_PM_S2RAM_CUSTOM_MARKING=y
  * CONFIG_PM_DEVICE=y
  * CONFIG_PM_DEVICE_RUNTIME=y

Set these Kconfigs as follows for the application running on the radio core:

  * CONFIG_PM=y
  * CONFIG_POWEROFF=y
  * CONFIG_PM_DEVICE=y
  * CONFIG_PM_DEVICE_RUNTIME=y

Consider also the following:

  * Disable all unused peripherals before entering sleep (Zephyr's API does this automatically when supported).
  * Add `zephyr,pm-device-runtime-auto` in the DTS for all peripherals with runtime PM support.
    ## DTS to JSON update here
  * Build and program an empty image on any unused core to release shared resources.
    ##sample of this empty image?

Deep-sleep policy
-----------------

Minimum sleep durations:

  * Suspend-to-Idle: ≥ 1 ms
  * Suspend-to-RAM: ≥ 2 ms

Use policy locks to prevent premature deep-sleep entry:

.. code-block:: c

   /* In active mode: lock deep states */
   pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
   pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_RAM, PM_ALL_SUBSTATES);

   /* Before going to sleep: unlock */
   pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);

>    pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_RAM, PM_ALL_SUBSTATES);

Reviewer: This section doesn't provide any useful information for the application developer. What was the author's intention?

Reviewer 3: It should be explained why it is needed. Without proper explanation it indeed seems useless.

Memory and cache optimizations
------------------------------

>   * Relocate code and data to TCM to avoid waking MRAM.

Reviewer: ```suggestion
  * Relocate frequently accessed code and data to local RAM to avoid waking global domains, especially MRAM, which uses high power.
```

  * Profile L1/L2 cache usage to minimize MRAM accesses.

>   * Disable MRAM latency manager (`CONFIG_MRAM_LATENCY=n`) to power MRAM off when idle.

Reviewer: This seems counterintuitive. Isn't MRAM constantly powered on if the latency manager is unavailable?

Author: @David-Garcia-Polo @alstrzebonski  @zycz could you check this?

Peripheral and clock recommendations
------------------------------------

>   * Use slow-domain (13X) peripherals to avoid waking the fast domain.

Reviewer: ```suggestion
  * Use global peripherals from the slow domain (peripheral index 13X) to avoid waking up the fast global domain, which uses more power.
```

  * If radio is active frequently, keep HFXO enabled (wake-up ≈ 800 µs).

>   * For SPI, consider nRFy or nrfx libraries instead of Zephyr’s blocking driver to save ~30 µA.

Reviewer: Is nrfy recommended to be used by an application developer? I thought they should stick to nrfx

Single-core vs. dual-core
-------------------------

  * A single-core solution (radio core only) can reduce IPC overhead (~0.15 mA at 1 kHz).
  * Larger TCM on the radio core allows more code to run off-MRAM.

>   * Always load an empty image on the unused core to free global resources.

Reviewer: ```suggestion
  * Always load an empty image on the unused core to free global resources allocated to this core by the System Controller on initialization.
```

* Sleep currents (5 V supply):

  * Application core (Suspend-to-RAM): < 10 µA
  * Radio core (Suspend-to-Idle, cache disabled): < 10 µA

* Active currents (estimates at 5 V):

  * Radio core in idle: ~0.25 mA
  * App core in idle: ~0.15 mA
  * Core wake-up every 1 ms: +0.05 mA
  * ADC sample every 1 ms: +0.10 mA
  * SPI transaction every 1 ms: +0.20 mA
  * IPC message every 1 ms: +0.15 mA

Recommended Kconfig configuration
=================================

Set these Kconfigs as follows for the application running on the application core:

  * CONFIG_PM=y
  * CONFIG_PM_S2RAM=y
  * CONFIG_POWEROFF=y
  * CONFIG_PM_S2RAM_CUSTOM_MARKING=y
  * CONFIG_PM_DEVICE=y
  * CONFIG_PM_DEVICE_RUNTIME=y

Set these Kconfigs as follows for the application running on the radio core:

  * CONFIG_PM=y
  * CONFIG_POWEROFF=y
  * CONFIG_PM_DEVICE=y
  * CONFIG_PM_DEVICE_RUNTIME=y

Consider also the following:

  * Disable all unused peripherals before entering sleep (Zephyr's API does this automatically when supported).
  * Add `zephyr,pm-device-runtime-auto` in the DTS for all peripherals with runtime PM support.
    ## DTS to JSON update here

>   * Build and program an empty image on any unused core to release shared resources.

Author: Do we have a sample of this empty image? @alstrzebonski @zycz
